src/gc/gcinterface.h

```diff
@@ -0,0 +1,394 @@
+extern uint8_t* g_GCShadow;
+extern uint8_t* g_GCShadowEnd;
+// saves the g_lowest_address in between GCs to verify the consistency of the shadow segment
+extern uint8_t* g_shadow_lowest_address;
+#endif
+
+// For low memory notification from host
+extern int32_t g_bLowMemoryFromHost;
+
+// !!!!!!!!!!!!!!!!!!!!!!!
+// make sure you change the def in bcl\system\gc.cs 
+// if you change this!
+enum collection_mode
+{
+    collection_non_blocking = 0x00000001,
+    collection_blocking = 0x00000002,
+    collection_optimized = 0x00000004,
+    collection_compacting = 0x00000008
+#ifdef STRESS_HEAP
+    , collection_gcstress = 0x80000000
+#endif // STRESS_HEAP
+};
+
+// !!!!!!!!!!!!!!!!!!!!!!!
+// make sure you change the def in bcl\system\gc.cs 
+// if you change this!
+enum wait_full_gc_status
+{
+    wait_full_gc_success = 0,
+    wait_full_gc_failed = 1,
+    wait_full_gc_cancelled = 2,
+    wait_full_gc_timeout = 3,
+    wait_full_gc_na = 4
+};
+
+// !!!!!!!!!!!!!!!!!!!!!!!
+// make sure you change the def in bcl\system\gc.cs 
+// if you change this!
+enum start_no_gc_region_status
+{
+    start_no_gc_success = 0,
+    start_no_gc_no_memory = 1,
+    start_no_gc_too_large = 2,
+    start_no_gc_in_progress = 3
+};
+
+enum end_no_gc_region_status
+{
+    end_no_gc_success = 0,
+    end_no_gc_not_in_progress = 1,
+    end_no_gc_induced = 2,
+    end_no_gc_alloc_exceeded = 3
+};
+
+typedef BOOL (* walk_fn)(Object*, void*);
+typedef void (* gen_walk_fn)(void *context, int generation, uint8_t *range_start, uint8_t * range_end, uint8_t *range_reserved);
+
+// IGCHeap is the interface that the VM will use when interacting with the GC.
+class IGCHeap {
+public:
+    virtual HRESULT Initialize() = 0;
+    virtual BOOL IsPromoted(Object *object) = 0;
+    virtual BOOL IsHeapPointer(void* object, BOOL small_heap_only = FALSE) = 0;
+    virtual Object* GetNextFinalizable() = 0;
+    virtual unsigned WhichGeneration(Object* obj) = 0;
+    virtual size_t GetTotalBytesInUse() = 0;
+    virtual unsigned GetCondemnedGeneration() = 0;
+    virtual int GetGcLatencyMode() = 0;
+    virtual int SetGcLatencyMode(int newLatencyMode) = 0;
+    virtual int GetLOHCompactionMode() = 0;
+    virtual void SetLOHCompactionMode(int newLOHCompactionMode) = 0;
+    virtual BOOL RegisterForFullGCNotification(uint32_t gen2Percentage, uint32_t lohPercentage) = 0;
+    virtual BOOL CancelFullGCNotification() = 0;
+    virtual int WaitForFullGCApproach(int millisecondsTimeout) = 0;
+    virtual int WaitForFullGCComplete(int millisecondsTimeout) = 0;
+    virtual int CollectionCount(int generation, int get_bgc_fgc_coutn = 0) = 0;
+    virtual size_t GetLastGCStartTime(int generation) = 0;
+    virtual size_t GetLastGCDuration(int generation) = 0;
+    virtual size_t GetNow() = 0;
+
+    // TODO(segilles) these are all referenced by the VM. Are all of them
+    // essential?
+    virtual BOOL    IsGCInProgressHelper (BOOL bConsiderGCStart = FALSE) = 0;
```

Comment:
most of them are - but we should group them and add some comments as our own notes about their usage (some APIs are only used by certainly components) so we can look them incrementally, eg,TemporaryEnable/DisableConcurrentGC is only used by profiling and we'll need to look at them again when we get to the profiling part.